homeassistant/components/panasonic_viera/media_player.py

```diff
@@ -149,98 +99,201 @@ def name(self):
+
+    async def async_play_media(self, media_type, media_id, **kwargs):
+        """Play media."""
+        await self._remote.async_play_media(media_type, media_id)
+
+
+class Remote:
+    """The Remote class. It stores the TV properties and the remote control connection itself."""
 
-    def media_pause(self):
-        """Send media pause command to media player."""
-        self._playing = False
-        self._remote.media_pause()
+    def __init__(
+        self, hass, host, port, app_id=None, encryption_key=None,
+    ):
+        """Initialize the Remote class."""
+        self._hass = hass
 
-    def media_next_track(self):
-        """Send next track command."""
-        self._remote.media_next_track()
+        self._host = host
+        self._port = port
+        self._app_id = app_id
+        self._encryption_key = encryption_key
+
+        self.state = None
+        self.volume = 0
+        self.muted = False
+        self.playing = True
 
-    def media_previous_track(self):
-        """Send the previous track command."""
-        self._remote.media_previous_track()
+        self._control = None
 
-    def play_media(self, media_type, media_id, **kwargs):
+    async def async_create_remote_control(self, during_setup=False):
+        """Create remote control."""
+        control_existed = self._control is not None
+        try:
+            params = {}
+            if self._app_id and self._encryption_key:
+                params["app_id"] = self._app_id
+                params["encryption_key"] = self._encryption_key
+
+            self._control = await self._hass.async_add_executor_job(
+                partial(RemoteControl, self._host, self._port, **params)
+            )
+
+            self.state = STATE_ON
+        except (TimeoutError, URLError, OSError) as err:
+            if control_existed or during_setup:
+                _LOGGER.error("Could not establish remote connection: %s", err)
+                self._control = None
+                self.state = STATE_OFF
+        except Exception as err:  # pylint: disable=broad-except
+            if control_existed or during_setup:
+                _LOGGER.error("An unknown error occurred: %s", err)
+                self._control = None
+                self.state = STATE_OFF
+
+    async def async_update(self):
+        """Retrieve the latest data."""
+        try:
+            if self._control is None:
+                raise Exception
+
+            self.muted = self._control.get_mute()
+            self.volume = self._control.get_volume() / 100
+            self.state = STATE_ON
+        except EncryptionRequired:
+            _LOGGER.error("The connection couldn't be encrypted.")
+        except Exception:  # pylint: disable=broad-except
+            self.state = STATE_OFF
+            await self.async_create_remote_control()
+
+    async def async_send_key(self, key):
+        """Send a key to the TV and handles exceptions."""
+        if self._control:
```

Comment:
Instead of having all these checks. Set the entity as unavailable to stop it from using services. Set the `available` property to False.

Reply:
actually, when I set `state` to `STATE_OFF`, the entity automatically becomes unavailable in the UI.  Can I leave it like that and remove the `if self._control:` checks?

Reply:
No we need to set it to unavailable to guard the services from being used.